src/proxy-ui-api/src/test/java/org/niis/xroad/restapi/auth/CsrfWebMvcTest.java

```diff
@@ -0,0 +1,181 @@
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
+@AutoConfigureMockMvc
+@Slf4j
+public class CsrfWebMvcTest {
+    public static final String XSRF_HEADER = "X-XSRF-TOKEN";
+    public static final String XSRF_COOKIE = "XSRF-TOKEN";
+    public static final String CSRF_PARAM = "_csrf";
+
+    private String username = "xroad-user";
+    private String tokenValue = "token";
+    private CsrfToken csrfToken = new DefaultCsrfToken(XSRF_HEADER, CSRF_PARAM, tokenValue);
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @SpyBean
+    @Qualifier(FORM_LOGIN_PAM_AUTHENTICATION)
+    private PamAuthenticationProvider pamAuthenticationProvider;
+
+    @SpyBean
+    private ApiKeyAuthenticationManager apiKeyAuthenticationManager;
+
+    @Before
+    public void setup() {
+        // setup mock auth in the SecurityContext and mock both auth providers (form login and api-key)
+        Authentication mockAuth = new UsernamePasswordAuthenticationToken(username, "pass");
+        SecurityContext securityContext = SecurityContextHolder.getContext();
+        securityContext.setAuthentication(mockAuth);
+        doReturn(mockAuth).when(pamAuthenticationProvider).authenticate(any());
+        doReturn(mockAuth).when(apiKeyAuthenticationManager).authenticate(any());
+    }
+
+    /**
+     * Test login with mocked authentication. Should return 200 with a valid CSRF token in a cookie
+     * @throws Exception
+     */
+    @Test
+    public void login() throws Exception {
+        MockHttpServletRequestBuilder mockRequest = MockMvcRequestBuilders.post("/login");
+        mockMvc.perform(mockRequest)
+                .andDo(print())
+                .andExpect(status().isOk())
+                .andExpect(cookie().exists(XSRF_COOKIE))
+                .andReturn()
+                .getResponse();
+    }
+
+    /**
+     * Test getting user data for the currently logged in user. Uses mocked authentication in a mock session.
+     * @throws Exception
+     */
+    @Test
+    public void getUser() throws Exception {
+        User expectedUser = new User().username(username);
+        String expectedUserJsonString = new Gson().toJson(expectedUser);
+        MockHttpServletRequestBuilder mockRequest = MockMvcRequestBuilders.get("/api/user")
+                .session(getMockSession())
+                .header(XSRF_HEADER, tokenValue)
+                .cookie(new Cookie(XSRF_COOKIE, tokenValue));
+        mockMvc.perform(mockRequest)
+                .andDo(print())
+                .andExpect(status().isOk())
+                .andExpect(content().json(expectedUserJsonString));
+    }
+
+    /**
+     * Test getting user data for the currently logged in user. Uses mocked authentication in a mock session.
+     * XSRF header value should not match
+     * @throws Exception
+     */
+    @Test
+    public void getUserCsrfHeaderFail() throws Exception {
```

Comment:
Nice to see you found a good way to implement these tests :+1: Might even be a topic worth of a short developer blog, informal "share the knowledge" video session, or whatever....